src/api-service/__app__/onefuzzlib/agent_events.py

```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/env python
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+import logging
+from typing import Optional, Union, cast
+from uuid import UUID
+
+from onefuzztypes.enums import (
+    ErrorCode,
+    NodeState,
+    NodeTaskState,
+    TaskDebugFlag,
+    TaskState,
+)
+from onefuzztypes.models import (
+    Error,
+    NodeDoneEventData,
+    NodeSettingUpEventData,
+    NodeStateUpdate,
+    WorkerDoneEvent,
+    WorkerEvent,
+    WorkerRunningEvent,
+)
+
+from ..onefuzzlib.pools import Node, NodeTasks
+from ..onefuzzlib.task_event import TaskEvent
+from ..onefuzzlib.tasks.main import Task
+
+
+def get_node(machine_id: UUID) -> Union[Node, Error]:
+    node = Node.get_by_machine_id(machine_id)
+    if not node:
+        return Error(code=ErrorCode.INVALID_NODE, errors=["unable to find node"])
+    return node
+
+
+def on_state_update(
+    machine_id: UUID,
+    state_update: NodeStateUpdate,
+) -> Optional[Error]:
+    state = state_update.state
+    node = get_node(machine_id)
+    if isinstance(node, Error):
+        return node
+
+    if state == NodeState.free:
+        if node.reimage_requested or node.delete_requested:
+            logging.info("stopping free node with reset flags: %s", node.machine_id)
+            node.stop()
+            return None
+
+        if node.could_shrink_scaleset():
+            logging.info("stopping free node to resize scaleset: %s", node.machine_id)
+            node.set_halt()
+            return None
+
+    if state == NodeState.init:
+        if node.delete_requested:
+            logging.info("stopping node (init and delete_requested): %s", machine_id)
+            node.stop()
+            return None
+        node.reimage_requested = False
+        node.state = state
+        node.save()
+        return None
+
+    logging.info("node state update: %s from:%s to:%s", machine_id, node.state, state)
+    node.state = state
+    node.save()
+
+    if state == NodeState.setting_up:
+        # Model-validated.
+        #
+        # This field will be required in the future.
+        # For now, it is optional for back compat.
+        setting_up_data = cast(
+            Optional[NodeSettingUpEventData],
+            state_update.data,
+        )
+
+        if setting_up_data:
+            if not setting_up_data.tasks:
+                raise Exception("setup without tasks.  machine_id: %s", machine_id)
```

Comment:
this should be an Error instance

Reply:
I used an exception on purpose here, as something is totally screwed up if we have a "setup tasks" with no tasks.

During our standup tomorrow, let's talk about this.  I think there is value in differentiating between errors that we should gracefully handle verses "ACK, THIS IS ALL WRONG" style issues.

Reply:
Moved to an error